test/unittest/base/test_datalayer.py

```diff
@@ -9,16 +9,136 @@
     torch = None
 
 
+import dataclasses as dc
+from unittest.mock import patch
+
 from superduperdb.backends.mongodb.query import Collection
+from superduperdb.base.artifact import Artifact
 from superduperdb.base.document import Document
 from superduperdb.base.exceptions import ComponentInUseError, ComponentInUseWarning
+from superduperdb.components.component import Component
 from superduperdb.components.dataset import Dataset
 from superduperdb.components.encoder import Encoder
 from superduperdb.components.listener import Listener
 
 n_data_points = 250
 
 
+@dc.dataclass
+class TestComponent(Component):
+    identifier: str
+    version: int = None
+    type_id = 'test-component'
+    is_on_create = False
+    is_on_load = False
+    is_schedule_jobs = False
+    child: 'TestComponent' = None
+    artifact: Artifact = None
+
+    def on_create(self, db):
+        self.is_on_create = True
+
+    def on_load(self, db):
+        self.is_on_load = True
+
+    def schedule_jobs(self, db, dependencies):
+        self.is_schedule_jobs = True
+        return []
+
+    @property
+    def child_components(self):
+        return [('child', 'test-component')] if self.child else []
+
+
+def test_add_version(local_empty_db):
+    # Check the component functions are called
+    component = TestComponent(identifier='test')
+    local_empty_db.add(component)
+    assert component.is_on_create is True
+    assert component.is_on_load is True
+    assert component.is_schedule_jobs is True
+    assert component.version == 0
+    assert local_empty_db.show('test-component', 'test') == [0]
+
+    # Test the component saves the data correctly
+    component_loaded = local_empty_db.load('test-component', 'test')
+    assert component.serialized == component_loaded.serialized
+
+    # Check duplicate components are not added
+    local_empty_db.add(component)
+    component_loaded = local_empty_db.load('test-component', 'test')
+    assert component_loaded.version == 0
+    assert local_empty_db.show('test-component', 'test') == [0]
+
+    # Check the version is incremented
+    component = TestComponent(identifier='test')
+    local_empty_db.add(component)
+    assert component.version == 1
```

Comment:
@blythed @jieguangzhou should the version be incremented even though same component is added 
although the instance are not same they are recreated!

Reply:
I think the version only can be auto incremented when `version==None`